### PR TITLE
Fixed #8936 -- Added view permissions and a read-only admin

### PR DESCRIPTION
**Pull request by PetrDlouhy**
This is implementation of ticket [8936](https://code.djangoproject.com/ticket/8936).
It adds 'view' permission to Django models options besides existing 'add', 'delete' and 'edit' permissions. It also changes the behaviour of admin if that permission is set. It completes the CRUD pattern for Django permissions.

The implementation is quite straightforward. Still, there few weak spots, that I am feeling that should be discussed with someone more experienced:
* Model situation: User has defined `ModelAdmin` with custom `has_change_permission()` such that it disallows to change (consequently also view) anything.
 * Then adding view permissions allows superuser to view that `ModelAdmin`.
 * I had to change `RowLevelChangePermissionModelAdmin` in tests for this.
* How should change `RelatedFieldWidgetWrapper`

### CODE REVIEW

**Comment by hobarrera** [comment]
LGTM. I like that you've been clear on the documentation on how this affects (or rather, doesn't) other permissions.
Last commit message has a typo; should read "forgotten".

**Comment by alytle** [comment]
Thanks for this, Petr. What's the typical timeline for changes like this to make stable?

**Comment by PetrDlouhy** [comment]
Thanks for the comments. I would like to get this task moving again. Can somebody tell, what I need to do next to get this merged?

@hobarrera: Message fixed

@alytle: I have no idea about the timeline.

**Comment by hobarrera** [comment]
Regarding time-line and getting it merged: Have you mentioned this on django-developers@ to get a bit more attention?

**Comment by PetrDlouhy** [comment]
Yes, here is the [thread](https://groups.google.com/forum/#!searchin/django-developers/view$20permissions/django-developers/X7YEGB9KJNc/IKiixDhECQAJ)

BTW: Do you know, why it failed on the `docs` task? I only changed the git commit comment. I don't see, what is wrong in the Jenkins log.

**Comment by jarshwah** [comment]
I've updated the status in trac (https://code.djangoproject.com/ticket/8936) to Ready for Checkin. This signals that the community has reviewed the patch, and it is ready for a core committer to take a look and possibly merge.

The docs task failing isn't your fault. It's a spelling mistake in an unrelated area - I received the same error in a PR of my own earlier today too.

**Comment by timgraham** [comment]
After a quick glance, it looks like release notes are missing. See also our [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist).

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: django/contrib/admin/options.py [hunk 1]
```diff
@@ -306,7 +306,16 @@ def get_readonly_fields(self, request, obj=None):
         """
         Hook for specifying custom readonly fields.
         """
-        return self.readonly_fields
+        if hasattr(request, 'user') and (self.has_change_permission(request, obj) or self.has_add_permission(request)):
```
Comment (timgraham) [inline django/contrib/admin/options.py]: Putting this logic in the default implementation of `get_readonly_fields()` could be problematic. Anyone overriding the default implementation will no longer be able to use the read-only admin, correct?

File: tests/admin_views/tests.py [hunk 2]
```diff
@@ -1694,7 +1740,7 @@ def test_change_view(self):
         r2 = RowLevelChangePermissionModel.objects.create(id=2, name="even id")
         change_url_1 = reverse('admin:admin_views_rowlevelchangepermissionmodel_change', args=(r1.pk,))
         change_url_2 = reverse('admin:admin_views_rowlevelchangepermissionmodel_change', args=(r2.pk,))
-        for login_dict in [self.super_login, self.changeuser_login, self.adduser_login, self.deleteuser_login]:
+        for login_dict in [self.super_login, self.viewuser_login, self.changeuser_login, self.adduser_login, self.deleteuser_login]:
```
Comment (timgraham) [inline tests/admin_views/tests.py]: please create the list separately so as to limit line lengths to 119 chars.
```
logins = [
    self.super_login,
    ...
]
```

File: tests/admin_views/tests.py [hunk 3]
```diff
@@ -1465,6 +1492,14 @@ def test_login(self):
         self.assertFalse(login.context)
         self.client.get(reverse('admin:logout'))
 
+        # View User
```
Comment (timgraham) [inline tests/admin_views/tests.py]: I'd prefer if you created separate test methods(e.g. `test_login_view_user`) for these additions as these test methods are already longer than normal.

File: docs/ref/contrib/admin/index.txt [hunk 4]
```diff
@@ -1691,6 +1691,16 @@ templates used by the :class:`ModelAdmin` views:
     Should return ``True`` if adding an object is permitted, ``False``
     otherwise.
 
+.. versionadded:: 1.9
```
Comment (timgraham) [inline docs/ref/contrib/admin/index.txt]: this annotation goes below `..method`

File: docs/ref/contrib/admin/index.txt [hunk 5]
```diff
@@ -1715,7 +1725,8 @@ templates used by the :class:`ModelAdmin` views:
     accessing the module's index page is permitted, ``False`` otherwise.
     Uses :meth:`User.has_module_perms()
     <django.contrib.auth.models.User.has_module_perms>` by default. Overriding
-    it does not restrict access to the add, change or delete views,
+    it does not restrict access to the view, add, change or delete views,
```
Comment (timgraham) [inline docs/ref/contrib/admin/index.txt]: please add a comma after "change"

File: docs/ref/models/options.txt [hunk 6]
```diff
@@ -283,7 +283,7 @@ Django quotes column and table names behind the scenes.
 
 .. attribute:: Options.default_permissions
 
-    Defaults to ``('add', 'change', 'delete')``. You may customize this list,
+    Defaults to ``('view', 'add', 'change', 'delete')``. You may customize this list,
```
Comment (timgraham) [inline docs/ref/models/options.txt]: need `.. versionchanged:: 1.9` annotation at the end of this section noting that the "view" permission was added.

**Comment by PetrDlouhy** [comment]
@timgraham Thank you very much for your review. I dealt with all your remarks and squashed the changes into one commit.

File: docs/releases/1.9.txt [hunk 7]
```diff
@@ -163,6 +163,8 @@ Minor features
 
 * JavaScript slug generation now supports Romanian characters.
 
+* Added ``view`` permission for read only access to the models.
```
Comment (timgraham) [inline docs/releases/1.9.txt]: This needs a more elaborate discussion of how the features works as well as links to the documentation added elsewhere.

File: django/contrib/admin/options.py [hunk 8]
```diff
@@ -308,6 +308,21 @@ def get_readonly_fields(self, request, obj=None):
         """
         return self.readonly_fields
 
+    def get_invariable_fields(self, request, obj=None):
+        """
+        Get fields that are either readonly or han't change/add permissions
```
Comment (timgraham) [inline django/contrib/admin/options.py]: han't -> typo?

File: django/contrib/admin/options.py [hunk 9]
```diff
@@ -308,6 +308,21 @@ def get_readonly_fields(self, request, obj=None):
...
+        """
+        if hasattr(request, 'user') and (self.has_change_permission(request, obj) or self.has_add_permission(request)):
```
Comment (timgraham) [inline django/contrib/admin/options.py]: please limit line length to 119 chars so horizontal scrolling isn't required to view the patch.

File: django/contrib/admin/options.py [hunk 10]
```diff
@@ -308,6 +308,21 @@ def get_readonly_fields(self, request, obj=None):
         """
         return self.readonly_fields
 
+    def get_invariable_fields(self, request, obj=None):
```
Comment (timgraham) [inline django/contrib/admin/options.py]: I don't know that "invariable" is the best word -- actually, I didn't know it was a word. I'll try to suggest something if you don't have any alternatives.

File: docs/ref/contrib/admin/index.txt [hunk 11]
```diff
@@ -1691,6 +1691,16 @@ templates used by the :class:`ModelAdmin` views:
     Should return ``True`` if adding an object is permitted, ``False``
     otherwise.
 
+.. method:: ModelAdmin.has_view_permission(request, obj=None)
+
+.. versionadded:: 1.9
```
Comment (timgraham) [inline docs/ref/contrib/admin/index.txt]: please indent this line

File: docs/ref/models/options.txt [hunk 12]
```diff
@@ -283,12 +283,16 @@ Django quotes column and table names behind the scenes.
...
     for example, by setting this to an empty list if your app doesn't require
     any of the default permissions. It must be specified on the model before
     the model is created by :djadmin:`migrate` in order to prevent any omitted
     permissions from being created.
 
+    .. versionchanged:: 1.9
+
+        The ``view`` permission was added
```
Comment (timgraham) [inline docs/ref/models/options.txt]: missing period

File: tests/admin_views/tests.py [hunk 13]
```diff
@@ -1480,6 +1507,24 @@ def test_login(self):
         form = login.context[0].get('form')
         self.assertEqual(form.errors['username'][0], 'This field is required.')
 
+    def test_login_view_user(self):
+        """
+        Make sure only staff members can log in.
```
Comment (timgraham) [inline tests/admin_views/tests.py]: This docstring should be customized for the purpose of this test (it looks to me like it's copied from the other test).

File: tests/admin_views/tests.py [hunk 14]
```diff
@@ -1974,6 +2051,18 @@ def test_has_module_permission(self):
...
+        """
+        Ensure that has_module_permission() returns True for all users who
+        have view permission for that module, so that
+        the module is displayed on the admin index page.
+        """
+        self.client.login(**self.viewuser_login)
+        response = self.client.get(self.index_url)
+        self.assertContains(response, 'admin_views')
+        self.assertContains(response, 'Articles')
+        self.client.logout()
```
Comment (timgraham) [inline tests/admin_views/tests.py]: no need to logout at the end of a test -- this is done automatically for each test

**Comment by timgraham** [comment]
After you update the patch, please also update the flags on the ticket so that it appears in the review queue.

**Comment by PetrDlouhy** [comment]
@timgraham Thanks. I processed your suggestions.
For the name of the function, I have used word "uneditable" now. If you have better suggestions, I can change it.

**Comment by timgraham** [comment]
I'm still missing discussion in the docs of how the permissions interact and what the upgrade considerations are. For example, the "change" permission seems to imply "view" permission.

**Comment by timgraham** [comment]
This also seems to need more checking around the edge cases. For example, If I can only view a model and its inlines and then add "delete" permission for an inline, the "delete" checkbox appears on the inline, but there isn't a submit button on the page.

**Comment by PetrDlouhy** [comment]
I am sorry, I accidentally deleted the original branch and can't reopen this PR even if I re-pushed the branch. So this continues at PR #5297

**Pull request closed without merging**